pkg/scheduler/queue/query_component_load_test.go

```diff
@@ -0,0 +1,217 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package queue
+
+import (
+	"math/rand"
+	"sync"
+	"testing"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/stretchr/testify/require"
+)
+
+// getForComponent is a test utility, not intended for use by consumers of QueryComponentLoad
+func (qcl *QueryComponentLoad) getForComponent(queryComponent QueryComponent) int {
+	qcl.inflightRequestsMu.RLock()
+	defer qcl.inflightRequestsMu.RUnlock()
+	return qcl.querierInflightRequestsByComponent[queryComponent]
+}
+
+func TestQueryComponentLoad_Concurrency(t *testing.T) {
+
+	requestCount := 100
+	testOverloadFactor := 2.0
+	queryComponentLoad, err := NewQueryComponentLoad(testOverloadFactor, prometheus.NewPedanticRegistry())
+	require.NoError(t, err)
+
+	mockForwardRequestToQuerier := func(t *testing.T, load *QueryComponentLoad) {
+		expectedQueryComponent := randAdditionalQueueDimension(false)[0]
+
+		load.IncrementForComponentName(expectedQueryComponent)
+		require.GreaterOrEqual(t, load.getForComponent(Ingester), 0)
+		require.GreaterOrEqual(t, load.getForComponent(StoreGateway), 0)
+
+		load.DecrementForComponentName(expectedQueryComponent)
+		require.GreaterOrEqual(t, load.getForComponent(Ingester), 0)
+		require.GreaterOrEqual(t, load.getForComponent(StoreGateway), 0)
+	}
+
+	wg := sync.WaitGroup{}
+	start := make(chan struct{})
+	for i := 0; i < requestCount; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			<-start
+			mockForwardRequestToQuerier(t, queryComponentLoad)
+		}()
+	}
+	close(start)
+	wg.Wait()
```

Comment:
Might be worth adding an assertion here to confirm that the inflight request count is 0 after all the incrementing and decrementing - this would give us extra confidence that there are no concurrency issues.